### PR TITLE
ARROW-13579: [C++] Expose Create EmptyArray, EmptyRecordBatch and EmptyTable utility functions.

### PR DESCRIPTION
**Pull request by ArianaVillegas**
Refactor CreateEmptyArray and CreateEmptyChunkedArray, and create CreateEmptyRecordBatch and CreateEmptyTable.

### CODE REVIEW

File: cpp/src/arrow/compute/kernels/util_internal.h [hunk 1]
```diff
@@ -23,8 +23,11 @@
 
 #include "arrow/array/util.h"
 #include "arrow/buffer.h"
+#include "arrow/chunked_array.h"
 #include "arrow/compute/kernels/codegen_internal.h"
 #include "arrow/compute/type_fwd.h"
+#include "arrow/record_batch.h"
+#include "arrow/table.h"
```
Comment (lidavidm) [inline cpp/src/arrow/compute/kernels/util_internal.h:30]: nit: I think we should be able to forward-declare Array/RecordBatch/Table instead of including the headers (or include `arrow/type_fwd.h`), which would cut down on the headers that get (transitively) included.

**Review by lidavidm — COMMENTED** [review]
LGTM. However, this doesn't appear to address the issue in the title, but only the related issue described in the JIRA comments. Perhaps we should split this off into its own JIRA?

**Review by aocsa — CHANGES_REQUESTED** [review]
This PR Looks GTM too. I suggestted some changes.  I agree with @lidavidm, to readdress this PR to Jira issue https://issues.apache.org/jira/browse/ARROW-13579 and after that Ariana can face issue https://issues.apache.org/jira/browse/ARROW-7051

File: cpp/src/arrow/compute/kernels/util_internal.h [hunk 2]
```diff
@@ -161,6 +164,21 @@ int64_t CopyNonNullValues(const Datum& datum, T* out) {
   return n;
 }
 
+//--------------------------------------------------------------------------
+// An internal function to create an empty array, chucked array, record batch and table.
+
+Result<std::shared_ptr<Array>> CreateEmptyArray(std::shared_ptr<DataType> type,
+                                                MemoryPool* memory_pool);
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/util_internal.h:171]: 1) Is there a reason these functions are in `compute/kernels/util_internal.h` rather than public APIs? These could go in the respective public headers, such as `array/util.h`, `chunked_array.h`, etc.

2) We generally use "Make" rather than "Create" when naming factory functions (e.g. "MakeArrayOfNull"), can you reuse the same convention here?

**Review by pitrou — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/compute/kernels/util_internal.h (hunk 2 above)
Comment (ArianaVillegas) [reply cpp/src/arrow/compute/kernels/util_internal.h:171]: 1. Not really. We can move these functions to their public APIs.
2. Yes, I can rename them. They would be MakeEmptyArray, MakeEmptyChukedArray, MakeEmptyRecordBatch and MakeEmptyTable.

File: cpp/src/arrow/compute/kernels/util_internal.h (hunk 1 above)
Comment (ArianaVillegas) [reply cpp/src/arrow/compute/kernels/util_internal.h:30]: I've already tried to include "arrow/type_fwd.h", but I got a message that says that classes have incomplete types and I can't use their methods.

Comment (lidavidm) [reply cpp/src/arrow/compute/kernels/util_internal.h:30]: Ah, sorry for the confusion - while, as suggested below, these functions will be moved, what would be done here is: 
1. `#include "arrow/type_fwd.h"` inside `util_internal.h`
2. `#include "arrow/record_batch.h"` etc. inside `util_internal.cc`

The motivation is just not to leak the includes to other users of the header, but of course in the implementation we still need the full include.

Comment (ArianaVillegas) [reply cpp/src/arrow/compute/kernels/util_internal.h:30]: Oh, I got it. I understand it better now, thanks.

**Comment by ArianaVillegas** [comment]
Hi, everyone. Thanks for your feedback. I updated this PR, and I think it's ready for final review/merging. Let me know if anything needs to be taken care of from my end.

**Review by aocsa — APPROVED** [review]
Thanks @ArianaVillegas, looks good to me.

File: cpp/src/arrow/compute/kernels/util_internal.h (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/compute/kernels/util_internal.h:30]: Can you remove these includes now? They're not needed anymore.

File: cpp/src/arrow/chunked_array.h [hunk 3]
```diff
@@ -89,6 +89,17 @@ class ARROW_EXPORT ChunkedArray {
...
 
+  /// \brief Create an empty ChunkedArray of a given type
+  ///
+  /// The output ChunkedArray will have one chunk with an empty
+  /// array of the given type.
+  ///
+  /// \param[in] type DataType of the empty ChunkedArray
+  /// \param[in] memory_pool MemoryPool pointer
+  /// \return the resulting ChunkedArray
+  static Result<std::shared_ptr<ChunkedArray>> MakeEmptyChunkedArray(
```
Comment (pitrou) [inline cpp/src/arrow/chunked_array.h:100]: Either make this a top-level `MakeEmptyChunkedArray` function, or rename it to `ChunkedArray::MakeEmpty`.

File: cpp/src/arrow/record_batch.h [hunk 4]
```diff
@@ -60,6 +60,17 @@ class ARROW_EXPORT RecordBatch {
...
+  /// \brief Create an empty RecordBatch of a given schema
+  ///
+  /// The output RecordBatch will be created with DataTypes from
+  /// the given schema.
+  ///
+  /// \param[in] schema Schema of the empty RecordBatch
+  /// \param[in] memory_pool MemoryPool pointer
+  /// \return the resulting RecordBatch
+  static Result<std::shared_ptr<RecordBatch>> MakeEmptyRecordBatch(
+      std::shared_ptr<Schema> schema, MemoryPool* memory_pool);
```
Comment (pitrou) [inline cpp/src/arrow/record_batch.h:72]: Same suggestion as for `ChunkedArray`: make it a top-level function, or rename it.

File: cpp/src/arrow/table.h [hunk 5]
```diff
@@ -63,6 +63,17 @@ class ARROW_EXPORT Table {
...
 
+  /// \brief Create an empty Table of a given schema
+  ///
+  /// The output Table will be created from a ChunkedArrayVector
+  /// with DataTypes from the schema.
+  ///
+  /// \param[in] schema Schema of the empty Table
+  /// \param[in] memory_pool MemoryPool pointer
+  /// \return the resulting Table
+  static Result<std::shared_ptr<Table>> MakeEmptyTable(std::shared_ptr<Schema> schema,
```
Comment (pitrou) [inline cpp/src/arrow/table.h:74]: And same suggestion here :-)

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks for the updates. The PR is lacking tests, can you add some?

**Comment by ArianaVillegas** [comment]
I've already made the requested changes. I think it's ready.

File: cpp/src/arrow/array/array_test.cc [hunk 6]
```diff
@@ -654,6 +654,44 @@ TEST_F(TestArray, TestMakeArrayFromMapScalar) {
...
+                                       uint64(),
+                                       float64(),
+                                       binary(),
+                                       large_binary(),
+                                       fixed_size_binary(3),
+                                       decimal(16, 4),
+                                       utf8(),
+                                       large_utf8(),
+                                       list(utf8()),
+                                       list(int64()),  // ARROW-9071
```
Comment (pitrou) [inline cpp/src/arrow/array/array_test.cc:676]: The comment seems irrelevant here, can you remove it?

File: cpp/src/arrow/array/util.h [hunk 7]
```diff
@@ -54,6 +54,17 @@ ARROW_EXPORT
 Result<std::shared_ptr<Array>> MakeArrayFromScalar(
     const Scalar& scalar, int64_t length, MemoryPool* pool = default_memory_pool());
 
+/// \brief Create an empty Array of a given type
+///
+/// The output Array will be of the given type.
+///
+/// \param[in] type DataType of the empty Array
+/// \param[in] pool MemoryPool pointer
```
Comment (pitrou) [inline cpp/src/arrow/array/util.h:62]: This is not really informative. Make this for example "the memory pool to allocate from", like above?

File: cpp/src/arrow/chunked_array.h [hunk 8]
```diff
@@ -89,6 +89,17 @@ class ARROW_EXPORT ChunkedArray {
   static Result<std::shared_ptr<ChunkedArray>> Make(
       ArrayVector chunks, std::shared_ptr<DataType> type = NULLPTR);
 
+  /// \brief Create an empty ChunkedArray of a given type
+  ///
+  /// The output ChunkedArray will have one chunk with an empty
+  /// array of the given type.
+  ///
+  /// \param[in] type DataType of the empty ChunkedArray
+  /// \param[in] pool MemoryPool pointer
```
Comment (pitrou) [inline cpp/src/arrow/chunked_array.h:98]: Same comment here.

File: cpp/src/arrow/chunked_array_test.cc [hunk 9]
```diff
@@ -68,6 +69,13 @@ TEST_F(TestChunkedArray, Make) {
   ASSERT_RAISES(Invalid, ChunkedArray::Make({chunk0}, int16()));
 }
 
+TEST_F(TestChunkedArray, MakeEmpty) {
+  ASSERT_OK_AND_ASSIGN(std::shared_ptr<ChunkedArray> empty,
+                       ChunkedArray::MakeEmpty(int64()));
+  AssertTypeEqual(*int64(), *empty->type());
+  ASSERT_EQ(empty->length(), 0);
```
Comment (pitrou) [inline cpp/src/arrow/chunked_array_test.cc:76]: Can you also add a validation test? Should probably be `ASSERT_OK(empty->ValidateFull())`.

File: cpp/src/arrow/compute/kernels/util_internal.h (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/compute/kernels/util_internal.h:29]: @ArianaVillegas Again, there doesn't seem to be a reason for adding these includes while the code didn't change.

File: cpp/src/arrow/compute/kernels/vector_selection.cc [hunk 10]
```diff
@@ -19,26 +19,21 @@
...
 #include "arrow/array/concatenate.h"
 #include "arrow/buffer_builder.h"
-#include "arrow/chunked_array.h"
 #include "arrow/compute/api_vector.h"
 #include "arrow/compute/kernels/common.h"
 #include "arrow/compute/kernels/util_internal.h"
 #include "arrow/extension_type.h"
-#include "arrow/record_batch.h"
 #include "arrow/result.h"
-#include "arrow/table.h"
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/vector_selection.cc:36]: I'm surprised those includes are not needed anymore. `vector_selection.cc` certainly processes tables, record batches and chunked arrays...

File: cpp/src/arrow/record_batch.h [hunk 11]
```diff
@@ -60,6 +60,17 @@ class ARROW_EXPORT RecordBatch {
       std::shared_ptr<Schema> schema, int64_t num_rows,
       std::vector<std::shared_ptr<ArrayData>> columns);
 
+  /// \brief Create an empty RecordBatch of a given schema
+  ///
+  /// The output RecordBatch will be created with DataTypes from
+  /// the given schema.
+  ///
+  /// \param[in] schema Schema of the empty RecordBatch
+  /// \param[in] pool MemoryPool pointer
```
Comment (pitrou) [inline cpp/src/arrow/record_batch.h:69]: Same comment here as well.

File: cpp/src/arrow/record_batch_test.cc [hunk 12]
```diff
@@ -317,4 +317,18 @@ TEST_F(TestRecordBatch, FromStructArrayInvalidNullCount) {
...
+  auto f1 = field("f1", uint8());
+  auto f2 = field("f2", int16());
+
+  std::vector<std::shared_ptr<Field>> fields = {f0, f1, f2};
+  auto schema = ::arrow::schema({f0, f1, f2});
+
+  ASSERT_OK_AND_ASSIGN(std::shared_ptr<RecordBatch> empty,
+                       RecordBatch::MakeEmpty(schema));
+  AssertSchemaEqual(*schema, *empty->schema());
+  ASSERT_EQ(empty->num_rows(), 0);
```
Comment (pitrou) [inline cpp/src/arrow/record_batch_test.cc:331]: Can you also validate the RecordBatch here, just like in the array tests?

File: cpp/src/arrow/table.h [hunk 13]
```diff
@@ -63,6 +63,17 @@ class ARROW_EXPORT Table {
                                      const std::vector<std::shared_ptr<Array>>& arrays,
                                      int64_t num_rows = -1);
 
+  /// \brief Create an empty Table of a given schema
+  ///
+  /// The output Table will be created from a ChunkedArrayVector
+  /// with DataTypes from the schema.
+  ///
+  /// \param[in] schema Schema of the empty Table
+  /// \param[in] pool MemoryPool pointer
```
Comment (pitrou) [inline cpp/src/arrow/table.h:72]: Same comment.

File: cpp/src/arrow/table.h [hunk 14]
```diff
@@ -63,6 +63,17 @@ class ARROW_EXPORT Table {
                                      const std::vector<std::shared_ptr<Array>>& arrays,
                                      int64_t num_rows = -1);
 
+  /// \brief Create an empty Table of a given schema
+  ///
+  /// The output Table will be created from a ChunkedArrayVector
+  /// with DataTypes from the schema.
```
Comment (pitrou) [inline cpp/src/arrow/table.h:69]: Hmm, the fact that a ChunkedArrayVector is used internally is irrelevant. Perhaps shorten this to:
"""The Table will be created with a single empty chunk per column""".

File: cpp/src/arrow/table_test.cc [hunk 15]
```diff
@@ -15,20 +15,21 @@
...
-#include <vector>
+#include "arrow/table.h"
 
 #include <gmock/gmock.h>
 #include <gtest/gtest.h>
 
+#include <cstdint>
+#include <memory>
+#include <vector>
+
```
Comment (pitrou) [inline cpp/src/arrow/table_test.cc:26]: I'm a bit surprised that you needed to reorder these includes, though this is not a problem. Out of curiosity, did an automatic formatter do this?

File: cpp/src/arrow/table_test.cc [hunk 16]
```diff
@@ -180,6 +181,19 @@ TEST_F(TestTable, Equals) {
...
+  auto f0 = field("f0", int32());
+  auto f1 = field("f1", uint8());
+  auto f2 = field("f2", int16());
+
+  std::vector<std::shared_ptr<Field>> fields = {f0, f1, f2};
+  auto schema = ::arrow::schema({f0, f1, f2});
+
+  ASSERT_OK_AND_ASSIGN(std::shared_ptr<Table> empty, Table::MakeEmpty(schema));
+  AssertSchemaEqual(*schema, *empty->schema());
+  ASSERT_EQ(empty->num_rows(), 0);
```
Comment (pitrou) [inline cpp/src/arrow/table_test.cc:194]: Can you add a validation check here as well?

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks for the update @ArianaVillegas . There are still a number of minor comments below, but we're nearing completion!

File: cpp/src/arrow/compute/kernels/vector_selection.cc (hunk 10 above)
Comment (ArianaVillegas) [reply cpp/src/arrow/compute/kernels/vector_selection.cc:36]: Well, as a previous suggestion I move the includes to util_internal, so in the future, we won't need to include RecordBatch/Table/ChukedArray files, just include util_internal.

Comment (pitrou) [reply cpp/src/arrow/compute/kernels/vector_selection.cc:36]: I don't see where such a suggestion was made. In general, we want to minimize the number of transitive includes in a `.h`, to try to control compile times.

File: cpp/src/arrow/table_test.cc (hunk 15 above)
Comment (ArianaVillegas) [reply cpp/src/arrow/table_test.cc:26]: Yes, I think the formatter just reorder it alphabetically.

**Comment by ArianaVillegas** [comment]
Requested changes are done :)

**Review by pitrou — APPROVED** [review]
